keps/sig-cli/0031-kustomize-integration.md

```diff
@@ -0,0 +1,230 @@
+    * [Non-Goals](#non-goals)
+* [Kustomize Introduction](#kustomize-introduction)    
+* [Proposal](#proposal)
+    * [UX](#UX)
+      * [apply](#apply)
+      * [get](#get)
+      * [delete](#delete)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Alternatives](#alternatives)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+[Kustomize](https://github.com/kubernetes-sigs/kustomize) is a tool developed to provide declarative support for kubernetes objects. It has been adopted by many projects and users. Having kustomize enabled in kubectl will address a list of long standing issues. This KEP describes how `kustomize` is integrated into kubectl subcommands with consistent UX.
+
+## Motivation
+
+Declarative specification of Kubernetes objects is the recommended way to manage Kubernetes applications or workloads. There is some gap in kubectl on declarative support. To eliminate the gap, a [KEP](https://github.com/kubernetes/community/blob/master/keps/sig-cli/0008-kustomize.md#faq) was proposed months ago and Kustomize was developed. After more than 10 iterations, Kustomize has a complete set of features and reached a good state to be integrated into kubectl. 
+
+### Goals
+
+Integrate kustomize with kubectl so that kubectl can recognize kustomization directories and expand resources from kustomization.yaml before running kubectl subcommands. This integration should be transparent. It doesn't change kubectl UX. This integration should also be backward compatible. For non kustomization directories, kubectl behaves the same as current. The integration shouldn't have any impact on those parts.
+
+
+### Non-Goals
+- provide an editing functionality of kustomization.yaml from kubectl
+- further integration with other kubectl flags
+
+## Kustomize Introduction
+
+Kustomize has following subcommands:
+- build
+- edit, edit also has subcommands
+   - Set 
+       - imagetag
+       - namespace
+       - nameprefix
+   - Add
+       - base
+       - resource
+       - patch
+       - label
+       - annotation
+       - configmap
+- config
+- version
+- help
+
+`edit` and `build` are most commonly used subcommands. 
+
+`edit` is to modify the fields in `kustomization.yaml`. A `kustomization.yaml` includes configurations that are consumed by Kustomize. Here is an example of `kustomization.yaml` file.
+
+`build` is to perform a set of pre-processing transformations on the resources inside one kustomization. Those transformations include:
+- Get objects from the base
+- Apply patches
+- Add name prefix to all resources
+- Add common label and annotation to all resources
+- Replace imageTag is specified
+- Update objects’ names where they are referenced
+- Resolve variables and substitute them
+
+```
+namePrefix: alices-
```

Comment:
How do you plan to version this config later and retain backward compatibility?

Reply:
`kustomization.yaml` has apiVersion, Kind fields, which will help versioning.

Reply:
So they are just omitted in this example?

Reply:
I don't see the fields in any examples:
https://github.com/kubernetes-sigs/kustomize/tree/master/examples

Reply:
They are omitted. Currently these two fields are optional in kustomization.yaml since kustomize has been backward compatible quite well. If there is any significant change in the future, we should require to set the version.

Reply:
What are the kind and version currently?

Reply:
not defined, which means users could use any version and kind and those two fields will be ignored.

Reply:
Please add a TODO to resolve that. Significant deviation from kubectl and Kubernetes behavior more generally needs to have strong justification, and adding a required version to an unversioned schema would be a non-backward-compatible change.

Reply:
Added a TODO for this. Also opened an [issue](https://github.com/kubernetes-sigs/kustomize/issues/588) in kustomize